File: graphql-api/src/main/java/io/prizy/publicapi/graphql/scalars/UUIDScalar.java
package io.prizy.publicapi.graphql.scalars;

import graphql.language.StringValue;
import graphql.schema.Coercing;
import graphql.schema.CoercingParseLiteralException;
import graphql.schema.CoercingParseValueException;
import graphql.schema.CoercingSerializeException;
import graphql.schema.GraphQLScalarType;
import lombok.experimental.UtilityClass;

import java.util.UUID;

/**
 * @author Nidhal Dogga
 * @created 6/24/2022 9:22 PM
 */


@UtilityClass
public class UUIDScalar {

  public GraphQLScalarType SCALAR = GraphQLScalarType.newScalar()
    .name("UUID")
    .description("UUID")
    .coercing(new InstantCoercing())
    .build();

  private class InstantCoercing implements Coercing<UUID, String> {

    @Override
    public String serialize(Object dataFetcherResult) throws CoercingSerializeException {
      if (dataFetcherResult instanceof UUID uuid) {
        return uuid.toString();
      }
      throw new CoercingSerializeException("Cannot serialize: " + dataFetcherResult);
    }

    @Override
    public UUID parseValue(Object input) throws CoercingParseValueException {
      return UUID.fromString(input.toString());
    }

    @Override
    public UUID parseLiteral(Object input) throws CoercingParseLiteralException {
      if (input instanceof StringValue stringValue) {
        return UUID.fromString(stringValue.getValue());
      }
      throw new CoercingParseLiteralException("Cannot parse value: " + input);
    }
  }

}

File: public-api/src/main/kotlin/io/prizy/publicapi/graphql/scalars/UUIDScalar.kt
package io.prizy.publicapi.graphql.scalars

import graphql.language.StringValue
import graphql.schema.Coercing
import graphql.schema.CoercingParseLiteralException
import graphql.schema.CoercingSerializeException
import graphql.schema.GraphQLScalarType
import java.util.UUID

object UUIDScalar {

  val graphqlScalarType: GraphQLScalarType = GraphQLScalarType.newScalar()
    .name("UUID")
    .description("UUID")
    .coercing(UUIDCoercing)
    .build()

  object UUIDCoercing : Coercing<UUID, String> {

    override fun serialize(dataFetcherResult: Any): String {
      when (dataFetcherResult) {
        is UUID -> return dataFetcherResult.toString()
        else -> throw CoercingSerializeException("Not a valid UUID")
      }
    }

    override fun parseValue(input: Any): UUID {
      return UUID.fromString(input.toString())
    }

    override fun parseLiteral(input: Any): UUID {
      when (input) {
        is StringValue -> return UUID.fromString(input.value)
        else -> throw CoercingParseLiteralException("Value is not a valid UUID")
      }
    }
  }
}